### PR TITLE
[GSoC][TMVA] Add PyTorch Tutorials

### PR DESCRIPTION
**Pull request by AnirudhDagar**
This PR aims to add tutorials of PyTorch Interface in TMVA and is an extension to the main PR i.e #5757. Feel free to comment and share feedback.

### CODE REVIEW

**Comment by AnirudhDagar** [comment]
I'll also share a few comaprision plots in this thread.

1. PyTorchClassification.py : TMVA_PyTorh_ROC

**Comment by AnirudhDagar** [comment]
@lmoneta This is now a complete set of tutorials both training and inference. It is ready for review.

**Comment by lmoneta** [comment]
It would be nice to have also a C++ version of the tutorial, maybe just one is enough (e.g. the classification one)
In that case it would be good to make cmake copying the pytorch python code from the test directory to the tutorial one too, as it is done for the test program

**Comment by lmoneta** [comment]
The tutorials work only with Python3 for an issue with the print statements.  
If possible, since ROOT still supports Python2, can we add support for both ?

**Comment by AnirudhDagar** [comment]
@lmoneta I added the `fstring` based prints because they are cleaner and officially the python community has discontinued python2 support since Jan 2020. I'll change those back to `format` based print strings to support python2 as well.

Also, the C tutorial is a good idea to have. I'll send in another PR to do that subsequently, once this is merged.

Thanks!

**Comment by AnirudhDagar** [comment]
@lmoneta the print statements now should be python2.7 compatible.

**Comment by omazapa** [comment]
Hi @AnirudhDagar  remember the c++ examples.


Best 
Omar.

**Comment by AnirudhDagar** [comment]
Extremely sorry for the reviews requested from all the code owners. It was automatically triggered due to a commit I pushed earlier.  Please ignore.

@lmoneta @omazapa I have added a CNN Tutorial in C for PyTorch. Please take a look.
Also, the tutorial is compatible with the GPU.

File: tutorials/tmva/TMVA_CNN_Classification.C [hunk 1]
```diff
@@ -461,6 +461,21 @@ void TMVA_CNN_Classification(std::vector<bool> opt = {1, 1, 1, 1})
       }
    }
 
+   if (usePyTorchCNN) {
```
Comment (lmoneta) [inline tutorials/tmva/TMVA_CNN_Classification.C:464]: the variable usePyTorchCNN is not defined. 
Please add this line at the beginning of the function TMVA_CNN_Classification:
```
 bool usePyTorchCNN = (opt.size() > 4) ? opt[4] : false; 
````

File: tutorials/tmva/TMVA_CNN_Classification.C [hunk 2]
```diff
@@ -461,6 +461,21 @@ void TMVA_CNN_Classification(std::vector<bool> opt = {1, 1, 1, 1})
...
+
+      Info("TMVA_CNN_Classification", "Using Convolutional PyTorch Model");
+      // execute
+      gSystem->Exec("python PyTorch_Generate_CNN_Model.py");
```
Comment (lmoneta) [inline tutorials/tmva/TMVA_CNN_Classification.C:468]: I would do a check if the file exists and modify the code as following:
```
   if (usePyTorchCNN) {

      Info("TMVA_CNN_Classification", "Using Convolutional PyTorch Model");
      // execute Python code for Pytorch
      TString pytorchFileName = "PyTorch_Generate_CNN_Model.py";
      if (gSystem->AccessPathName(pytorchFileName)) {
         Error("TMVA_CNN_Classification","Python file %s defining the PyTorch model is missing - skip using PyTorch",pytorchFileName.Data());
      }
      else {
         gSystem->Exec(TString::Format("python %s",pytorchFileName.Data()) );

         // book PyKeras method only if Keras model could be created
         Info("TMVA_CNN_Classification", "Booking PyTorch CNN model");
         factory.BookMethod(
            loader, TMVA::Types::kPyTorch, "PyTorch",
            "H:!V:VarTransform=None:FilenameModel=PyTorchModelCNN.pt:"
            "FilenameTrainedModel=PyTorchTrainedModelCNN.pt:NumEpochs=20:BatchSize=100"
            ":UserCode=PyTorch_Generate_CNN_Model.py");
      }
   }
```

Is it necessary to execute the python code `"PyTorch_Generate_CNN_Model.py` before in the C++ macro when we will also execute later in the MethodPyTorch class when we set up a model ? 

However after trying several executing the script from the MethodPyTorch class fails. Here is the error message:

```
Info in <TMVA_CNN_Classification>: Using Convolutional PyTorch Model
Info in <TMVA_CNN_Classification>: Booking PyTorch CNN model
Factory                  : Booking method: PyTorch
                         : 
                         : Using PyTorch - setting special configuration options 
                         : Using PyTorch version 1
                         :  Setup PyTorch Model 
                         :  Executing user initialization code from  PyTorch_Generate_CNN_Model.py
<WARNING>                : Failed to run python code: exec(open('PyTorch_Generate_CNN_Model.py').read())
<WARNING>                : Python error message:
Traceback (most recent call last):
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/_utils_internal.py", line 49, in get_source_lines_and_file
    sourcelines, file_lineno = inspect.getsourcelines(obj)
  File "/usr/lib/python3.6/inspect.py", line 955, in getsourcelines
    lines, lnum = findsource(object)
  File "/usr/lib/python3.6/inspect.py", line 786, in findsource
    raise OSError('could not get source code')
OSError: could not get source code

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 113, in <module>
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/_script.py", line 898, in script
    obj, torch.jit._recursive.infer_methods_to_compile
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 352, in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 406, in create_script_module_impl
    script_module = torch.jit.RecursiveScriptModule._construct(cpp_module, init_fn)
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/_script.py", line 388, in _construct
    init_fn(script_module)
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 388, in init_fn
    scripted = create_script_module_impl(orig_value, sub_concrete_type, stubs_fn)
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 364, in create_script_module_impl
    method_stubs = stubs_fn(nn_module)
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 591, in infer_methods_to_compile
    stubs.append(make_stub_from_method(nn_module, method))
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 52, in make_stub_from_method
    return make_stub(func, method_name)
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/_recursive.py", line 37, in make_stub
    ast = get_jit_def(func, name, self_name="RecursiveScriptModule")
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/jit/frontend.py", line 197, in get_jit_def
    sourcelines, file_lineno, filename = get_source_lines_and_file(fn, torch._C.ErrorReport.call_stack())
  File "/home/moneta/.local/lib/python3.6/site-packages/torch/_utils_internal.py", line 56, in get_source_lines_and_file
    raise OSError(msg) from e
OSError: Can't get source for <bound method Reshape.forward of Reshape()>. TorchScript requires source access in order to carry out compilation, make sure original .py files are available.
<FATAL>                         : Error executing the provided user code
```
@AnirudhDagar  do you know how to fix this ?

**Review by lmoneta — COMMENTED** [review]
The macro TMVA_CNN_Classification does not work with Pytorch enabled. 
It seems there is an error using torchgit  when running the code from C++.

File: tutorials/tmva/TMVA_CNN_Classification.C (hunk 2 above)
Comment (AnirudhDagar) [reply tutorials/tmva/TMVA_CNN_Classification.C:468]: @omazapa thanks for the fix. This should be resolved with your PR.

**Comment by omazapa** [comment]
Closing this, moved to https://github.com/root-project/root/pull/6935 with all the fixes.

**Pull request closed without merging**